### ISSUE TITLE
OptStackOps optimization case bypasses comparison generating incorrect code

### ISSUE THREAD

**Issue by bbbradsmith** [cc65/cc65#1989]
In the following example, the "fail" variation produces incorrect code with -O. This appears to be a regression since the 2.19 release.

([godbolt](https://godbolt.org/z/r86be4MWE))
```
unsigned char i,r,j;
void fail()
{
    if ((i & 0x1f) == (r ? 0 : 16)) j -=8;
}
void pass()
{
    if ((i & 0x1f) == (unsigned char)(r ? 0 : 16)) j -= 8;
}
```

Produces:
```
.proc   _fail: near
        lda     _i
        and     #$1F
        sta     ptr1
        lda     _r
        beq     L0005
        lda     #$00
        jmp     L0004
L0005:  lda     #$10
        cmp     ptr1
L0004:  bne     L0002
        lda     _j
        sec
        sbc     #$08
        sta     _j
L0002:  rts

.proc   _pass: near
        lda     _i
        and     #$1F
        sta     ptr1
        lda     _r
        beq     L0005
        lda     #$00
        jmp     L0006
L0005:  lda     #$10
L0006:  cmp     ptr1
        bne     L0002
        lda     _j
        sec
        sbc     #$08
        sta     _j
L0002:  rts
```

Note in the failure case how "jmp L0004" jumps over "cmp ptr1" instead of to it. A is never compared to ptr1 and instead the Z flag will be the result of "lda #$00" which causes it to branch never (i.e. the == acts as if always true instead of performing its intended comparison).

**Comment by bbbradsmith** [comment]
With --debug-opt-output it seems to fail at the first optimization rule applied: OptStackOps

```
=========================================================================
Initial code for function 'fail':
.segment	"CODE"

	ldx     #$00
	lda     _i
	ldx     #$00
	and     #$1F
	jsr     pushax
	lda     _r
	jeq     L0003
	ldx     #$00
	lda     #$00
	jmp     L0004
L0003:	ldx     #$00
	lda     #$10
L0004:	jsr     toseqax
	jeq     L0002
	ldx     #$00
	lda     _j
	sec
	sbc     #$08
	sta     _j
L0002:	rts

=========================================================================
Code after applying 'OptStackOps':
.segment	"CODE"

	ldx     #$00
	lda     _i
	ldx     #$00
	and     #$1F
	sta     ptr1
	lda     _r
	jeq     L0003
	ldx     #$00
	lda     #$00
	jmp     L0004
L0003:	ldx     #$00
	lda     #$10
	cmp     ptr1
	jsr     booleq
L0004:	jeq     L0002
	ldx     #$00
	lda     _j
	sec
	sbc     #$08
	sta     _j
L0002:	rts
```

**Comment by bbbradsmith** [comment]
For comparison, the passing version. Note that the intial code difference is just a redundant ldx #0, which will get optimized out in a later step, but is preventing whatever the bug in OptStackOps is.
```
=========================================================================
Initial code for function 'pass':
.segment	"CODE"

	ldx     #$00
	lda     _i
	ldx     #$00
	and     #$1F
	jsr     pushax
	lda     _r
	jeq     L0003
	ldx     #$00
	lda     #$00
	jmp     L0004
L0003:	ldx     #$00
	lda     #$10
L0004:	ldx     #$00
	jsr     toseqax
	jeq     L0002
	ldx     #$00
	lda     _j
	sec
	sbc     #$08
	sta     _j
L0002:	rts

=========================================================================
Code after applying 'OptStackOps':
.segment	"CODE"

	ldx     #$00
	lda     _i
	ldx     #$00
	and     #$1F
	sta     ptr1
	lda     _r
	jeq     L0003
	ldx     #$00
	lda     #$00
	jmp     L0004
L0003:	ldx     #$00
	lda     #$10
L0004:	ldx     #$00
	cmp     ptr1
	jsr     booleq
	jeq     L0002
	ldx     #$00
	lda     _j
	sec
	sbc     #$08
	sta     _j
L0002:	rts
```

**Comment by mrdudz** [comment]
It would be of great help if you could do some bisecting and find out what commit broke it... 2.9 was a LONG time ago :)

**Comment by bbbradsmith** [comment]
So, the problem with this specific case seems to be right here in _Opt_a_toscmpbool_: https://github.com/cc65/cc65/blob/c097401f8bd907b3994b68e7e8b47d50b4a1864c/src/cc65/coptstop.c#L1154

At the start of the function _D->IP_ is saves as an insertion point for the replaced boolean operator call (_jsr tosequax_).

In one of the branches, _AddStoreLhsA_ is used to insert a replacement for _jsr pushax_, which causes the index of the operator call to move forward, but neglects to update _D->IP_ to reflect this. This means that the replacement for the operator gets inserted before the line it was replacing instead of after, and when the original operator call is deleted its attached labels migrate downward to the code that followed it instead of before.

So, I have a simple fix that does that after _AddStoreLhsA_ (will submit a pull request once it satisfies _make test_ locally). I'm confident that it's an appropriate fix for this specific case.

I looked around to see if anything else looks like it might have suspicious activity. In most cases _D->IP_ seems to be used without branching before its use, but _Opt_a_tosicmp_ has a case which probably warrants a close look. Right here, we have an insertion at RHS which probably deserves the same fix, but I'm not sure how to create a test case that generates a label at an offending point: https://github.com/cc65/cc65/blob/c097401f8bd907b3994b68e7e8b47d50b4a1864c/src/cc65/coptstop.c#L1206

**Comment by bbbradsmith** [comment]
> It would be of great help if you could do some bisecting and find out what commit broke it... 2.9 was a LONG time ago :)

I believe this bug has existed since just before 2.19 was released. The offending code seems to have come from #1239 but it's hard to tell because that commit changes almost the whole file at once.

I'm not really sure when the test case we found would begin to trigger this bug, however.

**Comment by mrdudz** [comment]
fixed via #1992
